[fix] startpage engine / modified API

return42 opened the pull request:
## What does this PR do?

- [fix] startpage engine - avoid captcha
- [pylint] Startpage engine

## Why is this change important?

Startpage has introduced new anti-scraping measures that make SearXNG instances run into captchas:

1. some arguments has been removed and a new `sc` has been added.
2. search path changed from `do/search` to `sp/search`
3. POST request is no longer needed

## How to test this PR locally?

    make run

try to query `!sp foo`

## Author's checklist

## Related issues

Closes: https://github.com/searxng/searxng/issues/692

return42 committed:
[fix] startpage engine - avoid captcha

Startpage has introduced new anti-scraping measures that make SearXNG instances
run into captchas:

1. some arguments has been removed and a new `sc` has been added.
2. search path changed from `do/search` to `sp/search`
3. POST request is no longer needed

Closes: https://github.com/searxng/searxng/issues/692
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[pylint] Startpage engine

Fix remarks from pylint

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 commented on searx/engines/startpage.py:
```diff
@@ -46,24 +51,23 @@
-        'cmd': 'process_search',
-        'engine0': 'v1all',
+        # 'abp': "-1",
+        'sc': 'Mj4jZy61QETj20',
```
This `sc` argument is new.  I copied the value from an interactive startpage session.  I don't know what is encoded in the string.  This string might be invalid at one day.

EDIT: 

- In an interactive startpage session, the value of `sc` changes on each request (incl. _next_ page request).
- May be the IP of the SearXNG server, combined with a date signature is encoded in this string / we need some test from different IPs

not-my-profile replied:
Yeah, I think we should test this string again in 24 hours. We could also retrieve the `sc` string on startup by extracting it from the `hidden` input at https://startpage.com/ (requesting that page returns a new `sc` parameter each time).

dalf replied:
in case the `sc` parameter needs to be fetched: gigablast.py fetches the extra parameters every hour
https://github.com/searxng/searxng/blob/00d3a7477c01981a6ef2d396332670ad5b831fa6/searx/engines/gigablast.py#L81-L84

return42 replied:
I installed the patch on my server / lets see if this link will work in 24h https://darmarit.org/searx/search?q=%21sp%20Salvadore%20dali

return42 replied:
FYI `sc` is no longer valid / I will look for a solution like @dalf suggested in https://github.com/searxng/searxng/pull/695#discussion_r778790532

return42 replied:
To get an actual `sc` time stamp from startpage's home page I pushed b3ea061 .. from my local IP it works, but I did not have luck when I tested this patch on my server .. I have to dig a little bit deeper :-(

not-my-profile replied:
Maybe they already put your server IP on some kind of blacklist so that it doesn't get new `sc` values?

return42 replied:
This is what I also assume .. I will wait a day and see if it works, if not I have to analyze what the reason is.

Stratos42 replied:
> To get an actual `sc` time stamp from startpage's home page I pushed [b3ea061](https://github.com/searxng/searxng/commit/b3ea061a608af854614da067f83789f0001c68ff) .. from my local IP it works, but I did not have luck when I tested this patch on my server .. I have to dig a little bit deeper :-(

@return42 I just tried your patch on my server and it seems to work !

return42 replied:
@Stratos42 thanks for your feedback ... I did not have luck on my server, may be IP is blocked since a longer time .. may be something other went wrong .. before I merge this PR I will invest in more research .. 

Please keep me informed if the state of your server change / thanks!

return42 committed:
[fix] Get an actual `sc` argument from startpage's home page.

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[fix] startpage engine: fetch CAPTCHA & issues related to PR-695

In case of CAPTCHA raise a SearxEngineCaptchaException and suspend for 7 days.
When get_sc_code() fails raise a SearxEngineResponseException and suspend for 7
days.

[1] https://github.com/searxng/searxng/pull/695

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[mod] starpage engine: add comment about Startpage's FFox add-on

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

dalf committed:
[fix] startpage: workaround to use the startpage network

workaround for the issue #762

dalf reviewed (approved):
Thank you!
